src/session-pool.js

```diff
@@ -109,221 +97,71 @@ function SessionPool(database, options) {
 
   this.request_ = database.request;
   this.requestStream_ = database.requestStream;
-
-  this.reads_ = [];
-  this.writes_ = [];
-  this.borrowed_ = [];
-
-  this.pendingCreates_ = 0;
-  this.acquireQueue_ = new PQueue({concurrency: 1});
   this.requestQueue_ = new PQueue({concurrency: this.options.concurrency});
 
-  this.evictHandle_ = null;
-  this.pingHandle_ = null;
-
-  this.traces_ = new Map();
-
-  EventEmitter.call(this);
-}
-
-util.inherits(SessionPool, EventEmitter);
-
-/**
- * Formats stack trace objects into Node-like stack trace.
- *
- * @private
- *
- * @param {object[]} trace The trace object.
- * @return {string}
- */
-SessionPool.formatTrace_ = function(trace) {
-  var formatted = trace.slice(2).map(function(t) {
-    return `    at ${t.getFunctionName() ||
-      t.getMethodName()} (${t.getFileName()}:${t.getLineNumber()}:${t.getColumnNumber()})`;
-  });
-
-  return `Session leak detected!\n${formatted.join('\n')}`;
-};
-
-/**
- * Returns the number of available sessions.
- *
- * @return {number}
- */
-SessionPool.prototype.available = function() {
-  return this.reads_.length + this.writes_.length;
-};
-
-/**
- * Returns the number of borrowed sessions.
- *
- * @return {number}
- */
-SessionPool.prototype.borrowed = function() {
-  return this.borrowed_.length;
-};
-
-/**
- * Closes and the pool.
- *
- * @emits SessionPool#close
- * @returns {Promise}
- */
-SessionPool.prototype.close = function() {
-  var self = this;
-  var sessions = this.reads_.concat(this.writes_, this.borrowed_);
-
-  this.isOpen = false;
-
-  this.reads_ = [];
-  this.writes_ = [];
-  this.borrowed_ = [];
-
-  this.stopHouseKeeping_();
-
-  this.emit('empty');
-  this.emit('close');
-
-  return Promise.all(
-    sessions.map(function(session) {
-      return self.destroySession_(session);
-    })
+  this.readPool = Pool.createPool(
+    {
+      create: this.createReadSession.bind(this),
+      destroy: this.destroySession.bind(this),
+    },
+    {
+      max: this.options.maxReads,
```

Comment:
Since you use this options key twice, you should destructure it above and just use the variables, would be easier to check for valid properties and improve performances trivially